### PR TITLE
[SPARK-27610][YARN] Shade netty native libraries

### PR DESCRIPTION
**Pull request by amuraru**
## What changes were proposed in this pull request?

Fixed the `spark-<version>-yarn-shuffle.jar` artifact packaging to shade the native netty libraries:
- shade the `META-INF/native/libnetty_*` native libraries when packagin
the yarn shuffle service jar. This is required as netty library loader
derives that based on shaded package name.
- updated the `org/spark_project` shade package prefix to `org/sparkproject`
(i.e. removed underscore) as the former breaks the netty native lib loading.

This was causing the yarn external shuffle service to fail 
when spark.shuffle.io.mode=EPOLL

## How was this patch tested?
Manual tests

### CODE REVIEW

**Comment by vanzin** [comment]
How does this work? Won't the symbols in the native libraries still match the original, non-relocated names?

e.g. in the source for the native lib I see this:

```
static jint netty_epoll_native_JNI_OnLoad(JNIEnv* env, const char* packagePrefix) {
    // We must register the statically referenced methods first!
    if (netty_unix_util_register_natives(env,
            packagePrefix,
            "io/netty/channel/epoll/NativeStaticallyReferencedJniMethods",
            statically_referenced_fixed_method_table,
            statically_referenced_fixed_method_table_size) != 0) {
        return JNI_ERR;
    }
    // Register the methods which are not referenced by static member variables
    JNINativeMethod* dynamicMethods = createDynamicMethodsTable(packagePrefix);
    if (netty_unix_util_register_natives(env,
            packagePrefix,
            "io/netty/channel/epoll/Native",
            dynamicMethods,
            dynamicMethodsTableSize()) != 0) {
        freeDynamicMethodsTable(dynamicMethods);
        return JNI_ERR;
    }
```

The shade plugin doesn't change those strings to be in the new relocated package. Never mind how the JVM is even able to match the relocate Java names to the symbols in the native library in the first place.

**Comment by vanzin** [comment]
Hmmm... the netty native code seems to map the native symbols to java names based on the file name of the native library (see `netty_unix_util_parse_package_prefix`). Clever.

Usual requests:
- your PR title doesn't follow the convention (missing the [module] part)
- describe the fix, not the problem, in the PR title.

Also since it's a maven change you need to add `[test-maven]` to the PR title to run the build through maven.

**Comment by vanzin** [comment]
I'm still a little at a loss about how this works. Following the netty code, given the file name of the library, the translated name for e.g. `io/netty/channel/epoll/Native` will be `org/spark/project/io/netty/channel/epoll/Native`, because of this code:

```
    // Package names must be sanitized, in JNI packages names are separated by '/' characters.
    for (; temp != packageNameEnd; ++temp) {
        if (*temp == '_') {
            *temp = '/';
        }
    }
```

So the native library should fail to register because that class does not exist (see `netty_unix_util_register_natives`).

That shouldn't be hard to fix (just change the relocated name to not have an underscore), but it makes me wonder how you're verifying that this fix actually works.

**Comment by amuraru** [comment]
> That shouldn't be hard to fix (just change the relocated name to not have an underscore)

you're right, thanks for spotting that @vanzin  - I was testing with a different `spark.shade.packageName` value override in my environment

**Comment by amuraru** [comment]
> * describe the fix, not the problem, in the PR title.

I am not sure I get this, isn't the commit/PR title supposed to match JIRA title?
And JIRA title is filed as a bug, so problem is described in the title.
I find that more useful than say `Shade netty native libraries in yarn shuffle service uber jar`

File: common/network-yarn/pom.xml [hunk 1]
```diff
@@ -35,7 +35,7 @@
     <!-- Make sure all Hadoop dependencies are provided to avoid repackaging. -->
     <hadoop.deps.scope>provided</hadoop.deps.scope>
     <shuffle.jar>${project.build.directory}/scala-${scala.binary.version}/spark-${project.version}-yarn-shuffle.jar</shuffle.jar>
-    <shade>org/spark_project/</shade>
+    <shade>org/spark/project/</shade>
```
Comment (srowen) [inline common/network-yarn/pom.xml]: I think this was on purpose, to avoid packaging into the org.spark namespace

**Comment by vanzin** [comment]
> I am not sure I get this, isn't the commit/PR title supposed to match JIRA title?

No. Bugs explain problems. PRs explain fixes. If you want to know more about the problem, look at the bug.

File: common/network-yarn/pom.xml (hunk 1 above)
Comment (vanzin) [reply common/network-yarn/pom.xml]: Probably doesn't matter much anyway, but to be more consistent with the original intent, `org/sparkproject` would be better.

File: conf/log4j.properties.template [hunk 2]
```diff
@@ -28,8 +28,8 @@ log4j.appender.console.layout.ConversionPattern=%d{yy/MM/dd HH:mm:ss} %p %c{1}:
 log4j.logger.org.apache.spark.repl.Main=WARN
 
 # Settings to quiet third party logs that are too verbose
-log4j.logger.org.spark_project.jetty=WARN
-log4j.logger.org.spark_project.jetty.util.component.AbstractLifeCycle=ERROR
+log4j.logger.org.spark.project.jetty=WARN
```
Comment (vanzin) [inline conf/log4j.properties.template]: None of these other files should be changed.

File: pom.xml [hunk 3]
```diff
@@ -202,7 +202,7 @@
     <test.include.tags></test.include.tags>
 
     <!-- Package to use when relocating shaded classes. -->
-    <spark.shade.packageName>org.spark_project</spark.shade.packageName>
+    <spark.shade.packageName>org.spark.project</spark.shade.packageName>
```
Comment (vanzin) [inline pom.xml]: This one doesn't need to change.

File: common/network-yarn/pom.xml (hunk 1 above)
Comment (srowen) [reply common/network-yarn/pom.xml]: I recall we tried `spark-project` at some point and it was valid in bytecode but not in Java. `spark_project` ought to be a valid part of a package. Weak. OK I see the argument for `org.sparkproject`

Comment (vanzin) [reply common/network-yarn/pom.xml]: `spark_project` is valid, but you need to go and read the discussion to know why the change is needed. (Or the updated PR description which also mentions it.)

Comment (amuraru) [reply common/network-yarn/pom.xml]: Makes sense @srowen @vanzin - updated the patch to shade in `org.sparkproject`, ie no underscore

File: pom.xml (hunk 3 above)
Comment (amuraru) [reply pom.xml]: can you elaborate @vanzin  - `common/network-yarn` package uses this property from parent pom to get the shade prefix

File: conf/log4j.properties.template (hunk 2 above)
Comment (amuraru) [reply conf/log4j.properties.template]: Can you elaborate? `spark.shade.packageName` parent parameter is used across the entire project. 
Are you suggesting to have different shade prefix for `common/network-yarn` module only?

**Comment by amuraru** [comment]
> No. Bugs explain problems. PRs explain fixes. If you want to know more about the problem, look at the bug.

I see your point and the note in the Contributing Guide:
```The PR title should be of the form [SPARK-xxxx][COMPONENT] Title, where SPARK-xxxx is the relevant JIRA number, COMPONENT is one of the PR categories shown at spark-prs.appspot.com and Title may be the JIRA’s title or a more specific title describing the PR itself.```

I personally find it more consistent and easier to search (from a user perspective) the git history when the commit title describes the problem and not necessarily the fix, which can be described in the commit body.
Anyway, updated the commit and PR title per your suggestion.

File: pom.xml (hunk 3 above)
Comment (vanzin) [reply pom.xml]: Hmm, I see what the other one is doing now. My bad.

**Comment by vanzin** [comment]
> git history when the commit title describes the problem

At the risk of digressing too much, I find it extremely hard to understand a PR that explains what the problem is instead of explaining what it is doing. Many, many times the bug description is not a good way to understand the PR that fixes it.

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
ok to test

**Comment by vanzin** [comment]
It might be less work to override the shaded package name in the yarn shuffle service packaging...

**Comment by amuraru** [comment]
one last try to fix Mima.
@vanzin if this won't fix the issue - I can try to constrain the fix in the yarn package only

**Comment by vanzin** [comment]
One reason why I have a preference for the yarn-packaging-only fix is that this is potentially messing people's logging configuration, in case they saw a need to customize Jetty logs. I don't think that's a huge deal though.

Anyway, will leave it here until tomorrow to see if anyone else has comments.

**Comment by vanzin** [comment]
Merging to master.

**Pull request closed without merging**